src/esm_runscripts/workflow.py

```diff
@@ -36,253 +616,154 @@ def skip_cluster(cluster, config):
-            "workflow"
-        ]["first_task_in_queue"]
-        config["general"]["jobtype"] = config["general"]["workflow"][
-            "first_task_in_queue"
-        ]
-
-    return config
-
-
-def display_nicely(config):
-    esm_parser.pprint_config(config["general"]["workflow"])
-    return config
-
-
-def prepend_newrun_job(config):
-    gw_config = config["general"]["workflow"]
-    first_cluster_name = gw_config["first_task_in_queue"]
-    first_cluster = gw_config["subjob_clusters"][first_cluster_name]
-
-    if not first_cluster.get("batch_or_shell", "Error") == "SimulationSetup":
-
-        last_cluster_name = gw_config["last_task_in_queue"]
-        last_cluster = gw_config["subjob_clusters"][last_cluster_name]
-
-        new_first_cluster_name = "newrun"
-        new_first_cluster = {
-            "newrun": {
-                "called_from": last_cluster_name,
-                "run_before": first_cluster_name,
-                "next_submit": [first_cluster_name],
-                "subjobs": ["newrun_general"],
-                "batch_or_shell": "SimulationSetup",
-            }
-        }
-
-        last_cluster["next_submit"].append("newrun")
-        last_cluster["next_submit"].remove(first_cluster_name)
-
-        first_cluster["called_from"] = "newrun"
-
-        gw_config["first_task_in_queue"] = "newrun"
-
-        new_subjob = {
-            "newrun_general": {
-                "nproc": 1,
-                "called_from": last_cluster_name,
-                "run_before": first_cluster_name,
-                "next_submit": [first_cluster_name],
-                "subjob_cluster": "newrun",
-            }
-        }
-
-        gw_config["subjob_clusters"].update(new_first_cluster)
-        gw_config["subjobs"].update(new_subjob)
+class WorkflowPhase:
+    """A workflow phase class."""
+
+    def __init__(self, phase_name):
+        self.name = None
+        self.nproc = 1                              # needed
+        self.run_before = None
+        self.run_after = None
+        self.submit_to_batch_system = False         # needed
+        self.run_on_queue = None
+        self.cluster = None
+        self.next_submit = []                       # needed
+        self.called_from = None                     # needed
+        self.batch_or_shell = "SimulationSetup"     # needed
+        self.order_in_cluster = "sequential"        # needed ???
+        self.run_only = None
+        self.skip_chunk_number = None
+        self.skip_run_number = None
+        self.name = phase_name
+
+
+class UserWorkflowPhase(WorkflowPhase):
+    """A user workflow phase class."""
+
+    def __init__(self, phase_name):
+
+        WorkflowPhase.__init__(self, phase_name)
+
+        self.script = None
+        self.script_dir = None
+        self.call_function = None
+        self.env_preparation = None
+        self.batch_or_shell = "shell"
+        self.submit_to_batch_system = False
```

Comment:
IMO we don't need a UserWorkflowPhase, the user should be able to simply append to our default.